cpp/src/io/parquet/writer_impl.cu

```diff
@@ -792,6 +796,113 @@ void writer::impl::init_encoder_pages(hostdevice_2dvector<gpu::EncColumnChunk> &
+    print(d_vec, label);
+  }
+  template <typename T>
+  std::enable_if_t<!cudf::is_numeric<T>(), void> operator()(column_view const &col,
+                                                            std::string label = "")
+  {
+    CUDF_FAIL("no strings");
+  }
+};
+void print(column_view const &col, std::string label = "")
+{
+  cudf::type_dispatcher(col.type(), printer{}, col, label);
+}
+
+void snappy_compress(device_span<gpu_inflate_input_s> comp_in,
+                     device_span<gpu_inflate_status_s> comp_stat,
+                     rmm::cuda_stream_view stream)
+{
+  size_t num_comp_pages = comp_in.size();
+  do {
+    size_t temp_size;
+    nvcompError_t nvcomp_error =
+      nvcompBatchedSnappyCompressGetTempSize(num_comp_pages, 1 << 20, &temp_size);
+    if (nvcomp_error != nvcompError_t::nvcompSuccess) { break; }
+
+    // Not needed now but nvcomp API makes no promises about future
+    rmm::device_buffer scratch(temp_size, stream);
+    // Analogous to comp_in.srcDevice
+    rmm::device_uvector<void const *> uncompressed_data_ptrs(num_comp_pages, stream);
+    // Analogous to comp_in.srcSize
+    rmm::device_uvector<size_t> uncompressed_data_sizes(num_comp_pages, stream);
+    // Analogous to comp_in.dstDevice
+    rmm::device_uvector<void *> compressed_data_ptrs(num_comp_pages, stream);
+    // Analogous to comp_stat.bytes_written
+    rmm::device_vector<size_t> compressed_bytes_written(num_comp_pages);
+    // nvcomp does not currently use comp_in.dstSize. Cannot assume that the output will fit in
+    // the space allocated unless one uses the API nvcompBatchedSnappyCompressGetOutputSize()
+
+    // Prepare the vectors
+    auto comp_it = thrust::make_zip_iterator(uncompressed_data_ptrs.begin(),
+                                             uncompressed_data_sizes.begin(),
+                                             compressed_data_ptrs.begin());
+    thrust::transform(rmm::exec_policy(stream),
+                      comp_in.begin(),
+                      comp_in.end(),
+                      comp_it,
+                      [] __device__(gpu_inflate_input_s in) {
+                        return thrust::make_tuple(in.srcDevice, in.srcSize, in.dstDevice);
+                      });
+    nvcomp_error = nvcompBatchedSnappyCompressAsync(uncompressed_data_ptrs.data(),
+                                                    uncompressed_data_sizes.data(),
+                                                    num_comp_pages,
+                                                    scratch.data(),  // Not needed rn but future
+                                                    scratch.size(),
+                                                    compressed_data_ptrs.data(),
+                                                    compressed_bytes_written.data().get(),
+                                                    stream.value());
+
+    if (nvcomp_error != nvcompError_t::nvcompSuccess) { break; }
```

Comment:
IMO an exception would be better for error handling here. The `while(0)` loop looks odd to me.

Reply:
The while loop is a method to break out of execution but do something before returning. In this case, if nvcomp fails at any time, we set all the error codes per chunk to 1 to indicate compression failed for it. Works sort of like a goto. It's been used extensively in cub.

The purpose of using this instead of an exception is because we don't want to throw an exception in case compression fails. Instead, we'd continue writing uncompressed data. At least that's the old behaviour.